cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,241 @@
+   * {@code checkpoint.getNextEventTime()}.
+   * {@code -1} if no such offset can be found or {@code checkpoint.getNextOffset()} is negative.
+   */
+  long getMatchingOffset(final Checkpoint checkpoint, final int partition) throws IOException {
+    // No message should have next offset 0.
+    if (checkpoint.getNextOffset() == 0) {
+      return -1;
+    }
+    // Get BrokerInfo for constructing SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      throw new KafkaException(String.format("BrokerInfo from BrokerService is null for topic %s partition %d. " +
+                                               "Will retry in next run.",
+                                             topic, partition));
+    }
+    SimpleConsumer consumer
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT_MILLIS, BUFFER_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    try {
+      long timeStamp = getEventTimeByOffset(consumer, partition, clientId, checkpoint.getNextOffset() - 1);
+      if (timeStamp == checkpoint.getNextEventTime()) {
+        LOG.debug("Checkpoint {} in topic {} partition {} contains matching offset and timestamp. No need to search.",
+                  checkpoint, topic, partition);
+        return checkpoint.getNextOffset();
+      }
+    } catch (OffsetOutOfRangeException e) {
+      // Ignore OffsetOutOfRangeException
+    }
+
+    long nextOffset = findNextOffsetByTime(consumer, partition, clientId, checkpoint.getNextEventTime());
+    LOG.debug("Found new nextOffset {} for topic {} partition {} with existing checkpoint {}.",
+              nextOffset, topic, partition, checkpoint);
+    return nextOffset;
+  }
+
+  /**
+   * Determine the lower-bound and upper-bound of offsets to search for the next offset of the message
+   * with smallest offset and log event time equal to {@code targetTime}, and then perform the search.
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long findNextOffsetByTime(SimpleConsumer consumer, int partition, String clientName, long targetTime)
+    throws IOException {
+    // the offset of the earliest message
+    long minOffset;
+    try {
+      minOffset = KafkaUtil.getOffsetByTimestamp(consumer, topic, partition, targetTime);
+    } catch (IOException e) {
+      LOG.debug("Failed to find the segment with latest message published later than target time {}. " +
+                  "Find the offset with kafka.api.OffsetRequest.EarliestTime() instead.", targetTime, e);
+      minOffset = KafkaUtil.getOffsetByTimestamp(consumer, topic, partition, kafka.api.OffsetRequest.EarliestTime());
```

Comment:
`getOffsetByTimestamp` cannot return offset if it calls `getOffsetsBefore` and receives empty offset in the response. Therefore, If the `IOException` is thrown because targetTime is earlier than all existing Kafka messages, there's no point to retry with this targetTime again. If the `IOException` is because of Kafka problem, `KafkaUtil.getOffsetByTimestamp` inside catch block will throw it again.